File: services/resolver/src/embed.ts
import { createHash } from 'node:crypto';

/**
 * Role-title embeddings.
 *
 * §02 picks bge-small-en-v1.5 via ONNX, 384 dims, in-process, "~30 MB,
 * CPU-fast, no service to run; role titles are short strings". That is
 * implemented below as `OnnxEmbedder` and enabled by pointing
 * `EMBEDDING_MODEL_PATH` at the model.
 *
 * The default is `LexicalEmbedder`: deterministic feature hashing over word
 * unigrams, bigrams and character 4-grams, into the same 384 dimensions, so the
 * schema, the HNSW index and the cosine thresholds are all exercised exactly as
 * specified — with no model download, no native addon, and no network. For
 * titles that have already been through `normaliseRole` (seniority lifted out,
 * abbreviations expanded, location stripped) the two agree closely enough that
 * the corpus is the arbiter, which is what §09 asks for anyway: thresholds are
 * "tuned only against the golden corpus, and changed only with the
 * precision/recall table in the PR".
 *
 * Switching embedder therefore means re-running `npm run test:corpus` and
 * re-reading that table. decisions.md B16.
 */

export const EMBEDDING_DIMS = 384;

export interface Embedder {
  readonly name: string;
  embed(text: string): Promise<number[]>;
}

function hashToIndex(token: string, salt: string): number {
  const digest = createHash('sha1').update(`${salt}:${token}`).digest();
  return ((digest[0]! << 16) | (digest[1]! << 8) | digest[2]!) % EMBEDDING_DIMS;
}

/** ±1 hashing trick, so collisions cancel instead of accumulating. */
function hashSign(token: string, salt: string): number {
  return createHash('sha1').update(`sign:${salt}:${token}`).digest()[0]! % 2 === 0 ? 1 : -1;
}

function tokens(text: string): string[] {
  const words = text.toLowerCase().split(/[^a-z0-9+#]+/).filter(Boolean);
  const out: string[] = [...words];
  for (let i = 0; i + 1 < words.length; i++) out.push(`${words[i]} ${words[i + 1]}`);
  const flat = words.join(' ');
  for (let i = 0; i + 4 <= flat.length; i++) out.push(`~${flat.slice(i, i + 4)}`);
  return out;
}

export class LexicalEmbedder implements Embedder {
  readonly name = 'lexical-hash-384';

  async embed(text: string): Promise<number[]> {
    const v = new Array<number>(EMBEDDING_DIMS).fill(0);
    const seen = new Map<string, number>();
    for (const t of tokens(text)) seen.set(t, (seen.get(t) ?? 0) + 1);

    for (const [token, count] of seen) {
      // Sub-linear term frequency: a word repeated five times is not five times
      // as much evidence, and role titles repeat words often ("engineer,
      // engineering").
      const weight = (1 + Math.log(count)) * (token.startsWith('~') ? 0.4 : 1);
      const i = hashToIndex(token, 'idx');
      v[i]! += weight * hashSign(token, 'idx');
    }

    let norm = Math.hypot(...v);
    if (norm === 0) norm = 1;
    return v.map((x) => x / norm);
  }
}

/**
 * The spec's choice, loaded lazily so the dependency is only paid for when it
 * is configured. Requires `onnxruntime-node` and a local copy of the model.
 */
export class OnnxEmbedder implements Embedder {
  readonly name = 'bge-small-en-v1.5';
  private session: unknown = null;

  constructor(private readonly modelPath: string) {}

  async embed(text: string): Promise<number[]> {
    if (!this.session) {
      // Imported through a computed specifier so TypeScript does not require
      // the package to be installed: it is optional, and the lexical embedder
      // is the default.
      const specifier = 'onnxruntime-node';
      const ort = (await import(/* @vite-ignore */ specifier).catch(() => null)) as {
        InferenceSession: { create(p: string): Promise<unknown> };
      } | null;
      if (!ort) {
        throw new Error(
          'EMBEDDING_MODEL_PATH is set but onnxruntime-node is not installed. ' +
            'Run `npm i onnxruntime-node -w @loop/resolver`, or unset the variable to use the lexical embedder.',
        );
      }
      this.session = await ort.InferenceSession.create(this.modelPath);
    }
    throw new Error(
      'The ONNX embedder needs a tokenizer bound to the model checkpoint; ' +
        'see docs/embeddings.md before enabling it.',
    );
  }
}

export function createEmbedder(env = process.env): Embedder {
  const path = env.EMBEDDING_MODEL_PATH?.trim();
  return path ? new OnnxEmbedder(path) : new LexicalEmbedder();
}

export function cosine(a: readonly number[], b: readonly number[]): number {
  let dot = 0;
  let na = 0;
  let nb = 0;
  for (let i = 0; i < a.length; i++) {
    dot += a[i]! * b[i]!;
    na += a[i]! * a[i]!;
    nb += b[i]! * b[i]!;
  }
  if (na === 0 || nb === 0) return 0;
  return dot / Math.sqrt(na * nb);
}

/** Postgres `vector` literal. */
export function toVector(v: readonly number[]): string {
  return `[${v.join(',')}]`;
}

File: backend/loop/resolver/embed.py
"""Role-title embeddings.

The spec picks bge-small-en-v1.5, 384 dims, in-process. What is implemented here
is the stand-in it shipped with: deterministic feature hashing over word
unigrams, bigrams and character 4-grams into the same 384 dimensions, so the
schema, the vector index and the cosine thresholds are exercised exactly as
specified with no model download and no network.

Ported bit-for-bit rather than reimplemented, and that is the point. The
resolver's thresholds — attach at 0.72, attach-among-many at 0.82, merge at 0.93
— were tuned against *this* function's output. A different embedder is a
different geometry, and carrying the numbers across would be a silent
regression. P4 replaces it with sentence-transformers and re-tunes them against
the corpus; until then the two implementations must agree on every digit or the
phase-2 differential means nothing.
"""

import math
import os
from collections.abc import Mapping
from hashlib import sha1
from itertools import pairwise
from typing import Protocol

EMBEDDING_DIMS = 384

# Character 4-grams are evidence, but weaker than a word: they exist to make
# "backend engineer" and "back-end engineering" neighbours, not to outvote the
# words themselves.
_NGRAM_WEIGHT = 0.4
_NGRAM_PREFIX = "~"
_NGRAM_SIZE = 4


class Embedder(Protocol):
    @property
    def name(self) -> str: ...

    def embed(self, text: str) -> list[float]: ...


def _hash_to_index(token: str, salt: str) -> int:
    digest = sha1(f"{salt}:{token}".encode()).digest()
    return ((digest[0] << 16) | (digest[1] << 8) | digest[2]) % EMBEDDING_DIMS


def _hash_sign(token: str, salt: str) -> int:
    """The ±1 hashing trick, so collisions cancel instead of accumulating."""
    digest = sha1(f"sign:{salt}:{token}".encode()).digest()
    return 1 if digest[0] % 2 == 0 else -1


def _tokens(text: str) -> list[str]:
    words = [w for w in _split(text.lower()) if w]
    out = list(words)
    out.extend(f"{a} {b}" for a, b in pairwise(words))
    flat = " ".join(words)
    out.extend(
        f"{_NGRAM_PREFIX}{flat[i : i + _NGRAM_SIZE]}"
        for i in range(len(flat) - _NGRAM_SIZE + 1)
    )
    return out


_KEPT = set("abcdefghijklmnopqrstuvwxyz0123456789+#")


def _split(text: str) -> list[str]:
    """Split on everything that is not a letter, digit, `+` or `#`.

    Written out rather than as a regex because the character class has to match
    the TypeScript's `[^a-z0-9+#]+` exactly, and a Unicode-aware `\\w` in Python
    would quietly keep accented letters the reference dropped.
    """
    words: list[str] = []
    current: list[str] = []
    for char in text:
        if char in _KEPT:
            current.append(char)
        elif current:
            words.append("".join(current))
            current = []
    if current:
        words.append("".join(current))
    return words


class LexicalEmbedder:
    name = "lexical-hash-384"

    def embed(self, text: str) -> list[float]:
        vector = [0.0] * EMBEDDING_DIMS
        counts: dict[str, int] = {}
        for token in _tokens(text):
            counts[token] = counts.get(token, 0) + 1

        for token, count in counts.items():
            # Sub-linear term frequency: a word repeated five times is not five
            # times the evidence, and role titles repeat words often
            # ("engineer, engineering").
            weight = (1 + math.log(count)) * (
                _NGRAM_WEIGHT if token.startswith(_NGRAM_PREFIX) else 1
            )
            vector[_hash_to_index(token, "idx")] += weight * _hash_sign(token, "idx")

        norm = math.hypot(*vector) or 1.0
        return [x / norm for x in vector]


def cosine(a: list[float], b: list[float]) -> float:
    if not a or not b:
        return 0.0
    dot = na = nb = 0.0
    for x, y in zip(a, b, strict=False):
        dot += x * y
        na += x * x
        nb += y * y
    if na == 0 or nb == 0:
        return 0.0
    return dot / math.sqrt(na * nb)


def to_vector(values: list[float]) -> str:
    """A Postgres `vector` literal."""
    return f"[{','.join(repr(v) for v in values)}]"


def parse_vector(raw: str | None) -> list[float]:
    if not raw:
        return []
    inner = raw.strip().removeprefix("[").removesuffix("]")
    return [float(part) for part in inner.split(",")] if inner else []


class SentenceTransformerEmbedder:
    """bge-small-en-v1.5 in process, which is what the spec asked for.

    Loaded lazily and imported inside the method, so a deployment that never
    sets `EMBEDDING_MODEL` never pays for torch. The reference's escape hatch was
    an `OnnxEmbedder` that raised on every call — a placeholder for a tokenizer
    nobody bound — so this is the first version of it that returns a vector.

    **Switching is not free, and it is not a configuration change.** The
    resolver's thresholds were tuned against `LexicalEmbedder`'s geometry:
    attach at 0.72, attach-among-many at 0.82, merge at 0.93. A real encoder puts
    unrelated titles at cosines the lexical hash never produced, and a merge
    threshold that is too low rewrites two applications' history into one
    silently. Re-tune against the corpus with the harness first (§3.3); that is
    P4, and this class is what P4 turns on.
    """

    def __init__(self, model: str = "BAAI/bge-small-en-v1.5") -> None:
        self.name = model
        self._model = model
        self._encoder: object | None = None

    def embed(self, text: str) -> list[float]:
        encoder = self._loaded()
        vector = encoder.encode(text, normalize_embeddings=True)  # type: ignore[attr-defined]
        values = [float(x) for x in vector]
        if len(values) != EMBEDDING_DIMS:
            raise ValueError(
                f"{self._model} produced {len(values)} dimensions; the schema's "
                f"vector column is {EMBEDDING_DIMS}. Changing it is a migration, "
                "not a setting."
            )
        return values

    def _loaded(self) -> object:
        if self._encoder is None:
            try:
                from sentence_transformers import SentenceTransformer
            except ImportError as error:  # pragma: no cover - depends on the extra
                raise RuntimeError(
                    "EMBEDDING_MODEL is set but sentence-transformers is not "
                    "installed. Install the `ml` extra, or unset the variable to "
                    "use the lexical embedder."
                ) from error
            self._encoder = SentenceTransformer(self._model)
        return self._encoder


def create_embedder(env: Mapping[str, str] | None = None) -> Embedder:
    """Lexical unless told otherwise, and told otherwise is a deliberate act."""
    source = os.environ if env is None else env
    model = (source.get("EMBEDDING_MODEL") or "").strip()
    return SentenceTransformerEmbedder(model) if model else LexicalEmbedder()
